### PR TITLE
changes made during multisig integration push with lotus

### PR DESCRIPTION
**Pull request by whyrusleeping**
To be used in conjunction with: https://github.com/filecoin-project/lotus/pull/1148

### CODE REVIEW

File: actors/runtime/runtime.go [hunk 1]
```diff
@@ -170,3 +172,8 @@ type CBORMarshalable interface {
 type CBORUnmarshalable interface {
 	UnmarshalCBOR(r io.Reader) error
 }
+
+type CBORAble interface {
+	CBORMarshalable
+	CBORUnmarshalable
+}
```
Comment (frrist) [inline actors/runtime/runtime.go]: noting that here and elsewhere will conflict with https://github.com/filecoin-project/specs-actors/pull/45/files#diff-e2d8254b9ce8e65885afe1cf6a808a1dR173-R177

File: actors/builtin/multisig/multisig_actor.go [hunk 2]
```diff
@@ -16,9 +18,9 @@ type TxnID int64
 
 func (t TxnID) Key() string {
 	// convert a TxnID to a HAMT key.
-	txnKey := make([]byte, 0, binary.MaxVarintLen64)
-	binary.PutVarint(txnKey, int64(t))
-	return string(txnKey)
+	txnKey := make([]byte, binary.MaxVarintLen64)
+	n := binary.PutVarint(txnKey, int64(t))
+	return string(txnKey[:n])
```
Comment (frrist) [inline actors/builtin/multisig/multisig_actor.go]: :pray:

File: actors/builtin/multisig/multisig_actor.go [hunk 3]
```diff
@@ -69,12 +72,7 @@ type ProposeParams struct {
 	Params abi.MethodParams
 }
 
-type ProposeReturn struct {
-	TxnID TxnID
-}
-
-func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeReturn {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
+func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) *cbg.CborInt {
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Can we discuss this change?
1.  I added return structures symmetry with params objects, and consistent with service-method code generators everywhere (protobuf etc).
2. There's a CBOR-able big.Int type here already, we don't need to reference cbg directly here.

Comment (frrist) [inline actors/builtin/multisig/multisig_actor.go]: Why is this necessary? `ProposeReturn` was already CBORable.

File: actors/builtin/multisig/multisig_actor.go [hunk 4]
```diff
@@ -99,15 +97,15 @@ func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeR
...
+	out := cbg.CborInt(txnID)
+	return &out
 }
 
 type TxnIDParams struct {
 	ID TxnID
 }
 
 func (a *MultiSigActor) Approve(rt vmr.Runtime, params *TxnIDParams) *vmr.EmptyReturn {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: This is a significant item coming from the spec, the source already of much debate, and still here by fiat from jbenet. We can't just remove it, but your runtime shim can no-op it for now if you want.

File: actors/runtime/exitcode/reserved.go [hunk 5]
```diff
@@ -2,7 +2,7 @@ package exitcode
 
 import "strconv"
 
-type ExitCode int64
+type ExitCode uint8
```
Comment (anorth) [inline actors/runtime/exitcode/reserved.go]: We discussed this at length in person in November and since:
- we only use two integer types, with no unsigned variants
- there's no reason to restrict actors error code ranges, and if we did we can do it at runtime

Please revert this for now and cast in your shim

File: actors/builtin/multisig/multisig_actor.go (hunk 3 above)
Comment (whyrusleeping) [reply actors/builtin/multisig/multisig_actor.go]: It also serializes as a struct, causing extra wasted bytes, and deviates from existing expectations around return values

Comment (whyrusleeping) [reply actors/builtin/multisig/multisig_actor.go]: 1) Struct wrapping all return values feels wasteful to me, functions should be able to have simple return types, i'd rather not have to define a new type and run codegen for every new function that wants to return an int
2) This isnt a big int

File: actors/builtin/multisig/multisig_actor.go (hunk 4 above)
Comment (whyrusleeping) [reply actors/builtin/multisig/multisig_actor.go]: It's not doing anything productive. Its checking that the call is coming from some specific types of actors, when theres no good reason for that. So what if some new type of actor wants to interact with a multisig? The multisig explicitly doesnt care

File: actors/runtime/exitcode/reserved.go (hunk 5 above)
Comment (whyrusleeping) [reply actors/runtime/exitcode/reserved.go]: I don't recall these conversations, but we can make this change later, sure.

File: actors/builtin/multisig/multisig_actor.go (hunk 3 above)
Comment (anorth) [reply actors/builtin/multisig/multisig_actor.go]: Ok, I see the cost of the struct-wrapping bytes on chain. So methods can return serializable value, done.

And then I'm guessing CborInt is just a wrapper to implement the marshal methods around an int. cbg probably has a bunch of utilities like this that are unrelated to code generation but useful. Ok to land this now, but I might attempt to minimise external deps of this code later.

File: actors/builtin/multisig/multisig_actor.go (hunk 4 above)
Comment (anorth) [reply actors/builtin/multisig/multisig_actor.go]: I 100% agree with you but already lost this battle. Please restore it here for now so we can land these changes and then discuss elsewhere.

File: actors/runtime/exitcode/reserved.go (hunk 5 above)
Comment (anorth) [reply actors/runtime/exitcode/reserved.go]: Thanks!

File: actors/runtime/runtime.go (hunk 1 above)
Comment (anorth) [reply actors/runtime/runtime.go]: Yeah, sorry @whyrusleeping . I changed the interface names to match the more go-idiomatic ones from cbg.

File: actors/builtin/multisig/multisig_actor.go (hunk 4 above)
Comment (whyrusleeping) [reply actors/builtin/multisig/multisig_actor.go]: *sigh* fine

**Comment by whyrusleeping** [comment]
Alright, rebased and addressed comments. Should be good to go now

File: actors/builtin/multisig/multisig_actor.go [hunk 6]
```diff
@@ -10,15 +10,17 @@ import (
 	vmr "github.com/filecoin-project/specs-actors/actors/runtime"
 	exitcode "github.com/filecoin-project/specs-actors/actors/runtime/exitcode"
 	adt "github.com/filecoin-project/specs-actors/actors/util/adt"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: nit: prefer to group this with the other external import, `addr`
